### ISSUE TITLE
Knowing when a placeholder icon is returned

### ISSUE THREAD

**Issue by axxessgranted** [faviconkit/javascript-api#6]
Hi @AndreasPizsa :) 
Firstly, thanks for your work here!

I want to know if there is going to be any way to check whether the API returns the placeholder image (i.e grabbing a requested favicon was unsuccessful)? I don't see a way to do this due to 200s always being returned.

My use case: I want to display a custom icon in the case that grabbing the requested favicon was unsuccessful (instead of the placeholder icon that currently gets returned).

**Comment by andreaspitzer** [comment]
Hey @axxessgranted , thanks for your feedback and for your great question. I see where you’re coming from and your request makes a lot of sense.

It used to be that Favicon Kit would return an HTTP 404 for placeholder icons, with a placeholder icon in the response body. That worked well for most browsers, but Firefox would display a broken image icon instead, so we switched it back to always returning 200; in addition, Favicon Kit does return an `x-http-error-status` and `x-http-error-message` header when an error occurred.

Are you using Favicon Kit in a web app?

**Comment by axxessgranted** [comment]
Hi @AndreasPizsa , thanks for the prompt response!

I can definitely understand the decision given the pain that Firefox causes there. 

Yes, I'm using Favicon Kit in a React web app. Is there a way to leverage information from the headers to solve this issue, or do you have any other suggestion?

**Comment by andreaspitzer** [comment]
Hi again @axxessgranted ,
I understand your use case in general and I’ll take the time to think about it and come up with a good solution.

Can you help me better understand your specific needs?
+ What would your own placeholder images look like?
+ In what ways do the current placeholder images not meet your needs?
+ What would your ideal solution look like?

Thanks!

A.

**Comment by axxessgranted** [comment]
Hey @AndreasPizsa,

I'm using FaviconKit in a product, which of course has a theme designed by a design team. The design team has created custom placeholder icons which: 

- Fit the theme of the web app 
- Give users more information about the context of the target domain regarding specific settings that they can set inside the app. The same domain could have a different custom placeholder icon based on settings that the user has selected (it is up to the user to select target domains and choose how they are used in the app).

In this case, it is preferred that if the favicon is not grabbed, to rather display the custom icons which give some visual context of the use (or settings if you will) of the target domain within the app. 
So in short the placeholder images are contextual (which is why the current placeholder images do not meet my needs currently) and fitting the theme of the app. I hope this covers your first two questions 🙂 

My ideal solution: 
I would love to say return 404, but barring that possibility:
One way is perhaps to have a method, like `iconExists(iconUrl)` which will return true or false if a placeholder icon is going to be returned for the given iconUrl. Then we could know beforehand if a placeholder image is going to be returned and handle it accordingly.

I really appreciate you taking the time to consider this! Thank you!

**Comment by andreaspitzer** [comment]
@axxessgranted I’ll re-investigate options, inlcuding current handling of 404 response bodies in various browsers. If at all possible, returning a 404 with a default icon, but being able to override with an `onerror` handler seems like the best option to me, but I’ll know more after my investigation.

**Comment by andreaspitzer** [comment]
@axxessgranted Turns out Firefox 99 does correctly handle 404 errors with an image response body 👍  My next step is to see how far back, version-wise, this goes.

**Comment by andreaspitzer** [comment]
@axxessgranted A new version of Favicon Kit was just deployed.

For missing Favicons, Favicon Kit will now return a standard placeholder icon, as usual, and in addition also return a `404 Not Found` HTTP status code.

This enables you to add an `onerror` handler to your `<img>` tags and replace our standard placeholder with your own.

The change is effective immediately. Please note that for missing icons that have been requested just recently before the update, it can take up to 3 days for this change to become visible.

Can you please test the change and let me know if it’s working as expected? Thanks!

**Comment by axxessgranted** [comment]
Hi @AndreasPizsa,

Thank you! I have tested and this appears to work perfectly. I appreciate your efforts here and very quick response time, great work!
